src/main/java/org/opentripplanner/model/TimetableSnapshot.java

```diff
@@ -391,6 +371,46 @@ protected boolean clearRealtimeAddedTripPattern(String feedId) {
       );
   }
 
+  private void removeRealtimeUpdatedTripTimes(
+    TripPattern tripPattern,
+    FeedScopedId tripId,
+    LocalDate serviceDate
+  ) {
+    SortedSet<Timetable> sortedTimetables = this.timetables.get(tripPattern);
+    if (sortedTimetables != null) {
+      TripTimes tripTimesToRemove = null;
+      for (Timetable timetable : sortedTimetables) {
+        if (timetable.isValidFor(serviceDate)) {
+          final TripTimes tripTimes = timetable.getTripTimes(tripId);
+          if (tripTimes == null) {
+            LOG.debug("No triptimes to remove for trip {}", tripId);
+          } else if (tripTimesToRemove != null) {
+            LOG.debug("Found two triptimes to remove for trip {}", tripId);
+          } else {
+            tripTimesToRemove = tripTimes;
+          }
+        }
+      }
+
+      if (tripTimesToRemove != null) {
+        for (Timetable sortedTimetable : sortedTimetables) {
+          boolean isDirty = sortedTimetable.getTripTimes().remove(tripTimesToRemove);
+          if (isDirty) {
+            dirtyTimetables.add(sortedTimetable);
+          }
+        }
+      }
+    }
+  }
+
+  /**
+   * Removes the latest added trip pattern from the cache. This should be done when removing the
```

Comment:
Does this necessarily remove the trip pattern entirely from the cache? It seems to only remove the mapping from one particular TripIdAndServiceDate to the TripPattern, but the TripPattern could continue to be present for other keys.

The fact that this map is being referred to as a "cache" implies that the values are being reused, potentially for more than one tripId. I'm trying to establish whether there's an assumption throughout this part of the code that when a trip is changed such that it creates a new TripPattern, it's the only trip on that pattern.

Reply:
I think you are correct. I can improve javadoc.

Reply:
This is one of the things I want to change with refactoring of the transit model. The goal should be:

- One encapsulated call to add an entity (during graph build or RT). Then the one place will make sure the entity is inserted into the model and into any appropriate index. Since the one call probably should not know about indexes, we need to come up with a good design for this - alternatives exist.